runtime/interpreter/value.go

```diff
@@ -1570,13 +1571,17 @@ func (v *ArrayValue) IsImportable(inter *Interpreter) bool {
 }
 
 func (v *ArrayValue) checkInvalidatedResourceUse(interpreter *Interpreter, locationRange LocationRange) {
-	if v.isDestroyed || (v.array == nil && v.IsResourceKinded(interpreter)) {
+	if v.isDestroyed || v.IsStaleResource(interpreter) {
 		panic(InvalidatedResourceError{
 			LocationRange: locationRange,
 		})
 	}
 }
```

Comment:
seems like the definition of `checkInvalidatedResourceUse` is the same for all 3 types (`ArrayValue`, `CompositeValue`, `DictionaryValue`). this could be a free function or a method on `Interpreter`

Reply:
They are not, actually. for e.g: they rely on the atree value local field, which is different between, say `ArrayValue` and `CompositeValue`.

Reply:
ahh, fun stuff. could still potentially abstract it away with an anonymous interface (as ugly as that approach is)